src/rust/bitbox-aes/src/lib.rs

```diff
@@ -0,0 +1,3065 @@
+// Copyright 2025 Shift Crypto AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![no_std]
+
+#[cfg_attr(test, macro_use)]
+extern crate alloc;
+
+use alloc::vec::Vec;
+
+use hmac::{digest::FixedOutput, Mac, SimpleHmac};
+use sha2::Sha256;
+
+// AES block size.
+const BLOCK_SIZE: usize = 16;
+
+fn pad_length(buf_len: usize) -> usize {
+    BLOCK_SIZE - buf_len % BLOCK_SIZE
+}
+
+fn pkcs7_pad(buf: &mut Vec<u8>) {
+    let pad_len = pad_length(buf.len());
+    buf.resize(buf.len() + pad_len, pad_len as u8);
+}
+
+fn pkcs7_unpad(buf: &mut Vec<u8>) -> Result<(), ()> {
+    if buf.len() % BLOCK_SIZE != 0 {
+        return Err(());
+    }
+    let pad_len = *buf.last().ok_or(())? as usize;
```

Comment:
I don't understand `let pad_len = *buf.last().ok_or(())?`. Is pad_len the last in the cipher text?

Reply:
pkcs7 padding you fill the padding bytes with `pad_len`. To unpad, you read the pad_len and remove that many bytes.

Actually I could/should also check that the last `pad_len` bytes are all `pad_len` as a sanity check.

Reply:
Cool, I leave it up to you if you want to write that as a comment. I would have understood it without reading the pkcs7 padding spec if that was stated in a comment.

Reply:
Added the comments and the check that all `pad_len` bytes must equal `pad_len`.